# Ex - GroupBy

### Introduction:

GroupBy can be summarizes as Split-Apply-Combine.

Special thanks to: https://github.com/justmarkham for sharing the dataset and materials.

Check out this [Diagram](http://i.imgur.com/yjNkiwL.png)  
### Step 1. Import the necessary libraries

In [1]:
import pandas as pd

### Step 2. Import the dataset from drinks.csv

### Step 3. Assign it to a variable called drinks.

In [2]:
drinks = pd.read_csv('drinks.csv')

### Step 4. Which continent drinks more beer on average?

In [3]:
drinks.groupby('continent')['beer_servings'].mean()

continent
AF     61.471698
AS     37.045455
EU    193.777778
OC     89.687500
SA    175.083333
Name: beer_servings, dtype: float64

The continent which drinks more beers is Europe (EU).

### Step 5. For each continent print the statistics for wine consumption.

In [4]:
drinks.groupby('continent')['wine_servings'].describe()

count        mean        std  min   25%    50%     75%    max
continent                                                               
AF          53.0   16.264151  38.846419  0.0   1.0    2.0   13.00  233.0
AS          44.0    9.068182  21.667034  0.0   0.0    1.0    8.00  123.0
EU          45.0  142.222222  97.421738  0.0  59.0  128.0  195.00  370.0
OC          16.0   35.625000  64.555790  0.0   1.0    8.5   23.25  212.0
SA          12.0   62.416667  88.620189  1.0   3.0   12.0   98.50  221.0

Again, the continent where people drink more wine is Europe.

### Step 6. Print the mean alcohol consumption per continent for every column

In [5]:
drinks.groupby('continent').mean()

beer_servings  spirit_servings  wine_servings  \
continent                                                  
AF             61.471698        16.339623      16.264151   
AS             37.045455        60.840909       9.068182   
EU            193.777778       132.555556     142.222222   
OC             89.687500        58.437500      35.625000   
SA            175.083333       114.750000      62.416667   

           total_litres_of_pure_alcohol  
continent                                
AF                             3.007547  
AS                             2.170455  
EU                             8.617778  
OC                             3.381250  
SA                             6.308333

Overall, Europe is still being the continent which drinks more alcohol, probably because of Ireland (kidding!).

### Step 7. Print the median alcohol consumption per continent for every column

In [6]:
drinks.groupby('continent').median()

beer_servings  spirit_servings  wine_servings  \
continent                                                  
AF                  32.0              3.0            2.0   
AS                  17.5             16.0            1.0   
EU                 219.0            122.0          128.0   
OC                  52.5             37.0            8.5   
SA                 162.5            108.5           12.0   

           total_litres_of_pure_alcohol  
continent                                
AF                                 2.30  
AS                                 1.20  
EU                                10.00  
OC                                 1.75  
SA                                 6.85

If we take into account the median, Europe is still the winner. 

### Step 8. Print the mean, min and max values for spirit consumption.
#### This time output a DataFrame

In [7]:
drinks[['spirit_servings']].agg([pd.np.mean, pd.np.min, pd.np.max])

spirit_servings
mean        80.994819
amin         0.000000
amax       438.000000

In [8]:
drinks[drinks['spirit_servings'] == 438]

country  beer_servings  spirit_servings  wine_servings  \
68  Grenada            199              438             28   

    total_litres_of_pure_alcohol continent  
68                          11.9       NaN

There are some countries where they do not drink spirit drinks, whereas the maxium spirit cosumption is 438 servings which corresponds to the country of Grenada (surprisingly not in Europe). The average spirit consumtion is 80.99 servings.